File: src/api/routes/gameplay.py
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import desc, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from ...db.models import GamePlay
from ...db.session import get_db_session
from ..schemas import (
    GamePlayRankingItem,
    GamePlayRankingResponse,
    GamePlaySubmitRequest,
    GamePlaySubmitResponse,
)

router = APIRouter(prefix="/gameplay", tags=["gameplay"])


@router.post("", response_model=GamePlaySubmitResponse)
async def submit_gameplay(
    body: GamePlaySubmitRequest,
    db: AsyncSession = Depends(get_db_session),
) -> GamePlaySubmitResponse:
    """
    게임 플레이 데이터를 제출합니다.

    - **nickname**: 플레이어 닉네임
    - **score**: 최종 점수
    - **final_stage**: 도달한 최종 스테이지
    - **model_id**: 사용한 AI 모델 ID (beginner, intermediate, advanced)
    - **statistics**: 게임 통계 정보
    - **frames**: 프레임별 상세 데이터
    """
    try:
        # 프레임 데이터를 딕셔너리 리스트로 변환
        frames_data = [frame.model_dump() for frame in body.frames]

        # GamePlay 인스턴스 생성
        gameplay = GamePlay(
            nickname=body.nickname,
            score=body.score,
            final_stage=body.final_stage,
            model_id=body.model_id,
            total_frames=body.statistics.total_frames,
            play_duration=body.statistics.play_duration,
            enemies_destroyed=body.statistics.enemies_destroyed,
            shots_fired=body.statistics.shots_fired,
            hits=body.statistics.hits,
            deaths=body.statistics.deaths,
            frames_data=frames_data,
            created_at=datetime.now(tz=timezone.utc),
        )

        db.add(gameplay)
        await db.commit()
        await db.refresh(gameplay)

        return GamePlaySubmitResponse(
            id=gameplay.id,
            message="게임 플레이 데이터가 성공적으로 저장되었습니다.",
        )

    except Exception as e:
        await db.rollback()
        raise HTTPException(
            status_code=500,
            detail=f"게임 플레이 데이터 저장 중 오류가 발생했습니다: {str(e)}",
        )


@router.get("/rankings", response_model=GamePlayRankingResponse)
async def get_rankings(
    page: int = Query(1, ge=1, description="페이지 번호 (1부터 시작)"),
    page_size: int = Query(10, ge=1, le=100, description="페이지당 항목 수 (최대 100)"),
    model_id: Optional[str] = Query(
        None, description="특정 모델로 필터링 (예: beginner, intermediate, advanced)"
    ),
    db: AsyncSession = Depends(get_db_session),
) -> GamePlayRankingResponse:
    """
    게임 플레이 랭킹을 조회합니다.

    - **page**: 페이지 번호 (1부터 시작)
    - **page_size**: 페이지당 항목 수 (기본 10, 최대 100)
    - **model_id**: 특정 모델로 필터링 (선택 사항)

    점수 기준 내림차순으로 정렬되며, 동점일 경우 먼저 등록된 순서로 정렬됩니다.
    """
    try:
        # 기본 쿼리 구성 - 랭킹에 필요한 컬럼만 선택 (frames_data 제외)
        base_query = select(
            GamePlay.id,
            GamePlay.nickname,
            GamePlay.score,
            GamePlay.final_stage,
            GamePlay.model_id,
            GamePlay.total_frames,
            GamePlay.play_duration,
            GamePlay.created_at,
        )
        count_query = select(func.count(GamePlay.id))

        # 모델별 필터링
        if model_id:
            base_query = base_query.where(GamePlay.model_id == model_id)
            count_query = count_query.where(GamePlay.model_id == model_id)

        # 전체 개수 조회
        total_result = await db.execute(count_query)
        total = total_result.scalar_one()

        # 랭킹 조회 (점수 내림차순, 동점일 경우 생성일 오름차순)
        offset = (page - 1) * page_size
        rankings_query = (
            base_query.order_by(desc(GamePlay.score), GamePlay.created_at)
            .limit(page_size)
            .offset(offset)
        )

        result = await db.execute(rankings_query)
        rows = result.all()

        # 랭킹 아이템 생성
        rankings = [
            GamePlayRankingItem(
                id=row.id,
                nickname=row.nickname,
                score=row.score,
                final_stage=row.final_stage,
                model_id=row.model_id,
                total_frames=row.total_frames,
                play_duration=row.play_duration,
                created_at=row.created_at.isoformat() if row.created_at else "",
                rank=offset + idx + 1,
            )
            for idx, row in enumerate(rows)
        ]

        return GamePlayRankingResponse(
            rankings=rankings,
            total=total,
            page=page,
            page_size=page_size,
        )

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"랭킹 조회 중 오류가 발생했습니다: {str(e)}",
        )

File: worker/routes/gameplay.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { drizzle } from 'drizzle-orm/d1';
import { desc, eq, count } from 'drizzle-orm';
import { gameplays } from '../db/schema';
import { gamePlaySubmitRequestSchema } from '../lib/schemas';

type Env = {
  DB: D1Database;
};

const gameplayRouter = new Hono<{ Bindings: Env }>();

// 중복을 방지할 AI 에이전트 닉네임 목록
const AI_AGENT_NICKNAMES = ['beginner', 'medium', 'master'];

gameplayRouter.post('', zValidator('json', gamePlaySubmitRequestSchema), async (c) => {
  const body = c.req.valid('json');
  const db = drizzle(c.env.DB);
  
  try {
    // AI 에이전트 닉네임인 경우, 저장하지 않고 성공 응답 반환
    if (AI_AGENT_NICKNAMES.includes(body.nickname)) {
      return c.json({
        id: 'skipped-ai-agent',
        message: `AI 에이전트(${body.nickname})의 데이터는 저장되지 않았습니다.`,
      });
    }

    // 일반 유저인 경우 새로 삽입
    const result = await db.insert(gameplays).values({
      nickname: body.nickname,
      score: body.score,
      finalStage: body.final_stage,
      modelId: body.model_id,
      totalFrames: body.statistics.total_frames,
      playDuration: body.statistics.play_duration,
      enemiesDestroyed: body.statistics.enemies_destroyed,
      shotsFired: body.statistics.shots_fired,
      hits: body.statistics.hits,
      deaths: body.statistics.deaths,
      framesData: body.frames as any,
    }).returning();
    
    const gameplay = result[0];
    
    return c.json({
      id: gameplay.id,
      message: '게임 플레이 데이터가 성공적으로 저장되었습니다.',
    });
  } catch (error) {
    console.error('Error saving gameplay:', error);
    return c.json(
      {
        error: `게임 플레이 데이터 저장 중 오류가 발생했습니다: ${error instanceof Error ? error.message : String(error)}`,
      },
      500
    );
  }
});

gameplayRouter.get('/rankings', async (c) => {
  const db = drizzle(c.env.DB);
  
  // 쿼리 파라미터
  const page = parseInt(c.req.query('page') || '1', 10);
  const pageSize = Math.min(parseInt(c.req.query('page_size') || '10', 10), 100);
  const modelId = c.req.query('model_id');
  
  if (page < 1 || pageSize < 1) {
    return c.json({ error: 'Invalid page or page_size' }, 400);
  }
  
  try {
    // 조건 구성
    const conditions = modelId ? eq(gameplays.modelId, modelId) : undefined;
    
    // 전체 개수 조회
    const totalResult = await db
      .select({ count: count() })
      .from(gameplays)
      .where(conditions)
      .get();
    
    const total = totalResult?.count || 0;
    
    // 랭킹 조회
    const offset = (page - 1) * pageSize;
    const rankings = await db
      .select({
        id: gameplays.id,
        nickname: gameplays.nickname,
        score: gameplays.score,
        finalStage: gameplays.finalStage,
        modelId: gameplays.modelId,
        totalFrames: gameplays.totalFrames,
        playDuration: gameplays.playDuration,
        createdAt: gameplays.createdAt,
      })
      .from(gameplays)
      .where(conditions)
      .orderBy(desc(gameplays.score), gameplays.createdAt)
      .limit(pageSize)
      .offset(offset);
    
    // 랭킹 아이템 생성
    const rankingItems = rankings.map((row, idx) => ({
      id: row.id,
      nickname: row.nickname,
      score: row.score,
      final_stage: row.finalStage,
      model_id: row.modelId,
      total_frames: row.totalFrames,
      play_duration: row.playDuration,
      created_at: row.createdAt ? new Date(row.createdAt).toISOString() : '',
      rank: offset + idx + 1,
    }));
    
    return c.json({
      rankings: rankingItems,
      total,
      page,
      page_size: pageSize,
    });
  } catch (error) {
    console.error('Error fetching rankings:', error);
    return c.json(
      {
        error: `랭킹 조회 중 오류가 발생했습니다: ${error instanceof Error ? error.message : String(error)}`,
      },
      500
    );
  }
});

export default gameplayRouter;

